File: Roadmap/14 - FECHAS/typescript/nlarrea.ts
/*
 * EJERCICIO:
 * Crea dos variables utilizando los objetos fecha (date, o semejante) de tu lenguaje:
 * - Una primera que represente la fecha (día, mes, año, hora, minuto, segundo) actual.
 * - Una segunda que represente tu fecha de nacimiento (te puedes inventar la hora).
 * Calcula cuántos años han transcurrido entre ambas fechas.
 *
 * DIFICULTAD EXTRA (opcional):
 * Utilizando la fecha de tu cumpleaños, formatéala y muestra su resultado de
 * 10 maneras diferentes. Por ejemplo:
 * - Día, mes y año.
 * - Hora, minuto y segundo.
 * - Día de año.
 * - Día de la semana.
 * - Nombre del mes.
 * (lo que se te ocurra...)
 */


const currentDate: Date = new Date()
const birthdayDate: Date = new Date('1998-06-29 09:25:00')

const getYearsDifference = (dateOne: Date, dateTwo: Date): number => {
    if (dateOne > dateTwo) {
        return dateOne.getFullYear() - dateTwo.getFullYear()
    }

    return dateTwo.getFullYear() - dateOne.getFullYear()
}

console.log(getYearsDifference(currentDate, birthdayDate))


/*
 * DIFICULTAD EXTRA (opcional):
 * Utilizando la fecha de tu cumpleaños, formatéala y muestra su resultado de
 * 10 maneras diferentes. Por ejemplo:
 * - Día, mes y año.
 * - Hora, minuto y segundo.
 * - Día de año.
 * - Día de la semana.
 * - Nombre del mes.
 * (lo que se te ocurra...)
 */

const weekDays: Object = {
    0: 'Sunday',
    1: 'Monday',
    2: 'Tuesday',
    3: 'Wednesday',
    4: 'Thursday',
    5: 'Friday',
    6: 'Saturday'
}

const months: Object = {
    0: 'January',
    1: 'February',
    2: 'March',
    3: 'April',
    4: 'May',
    5: 'June',
    6: 'July',
    7: 'August',
    8: 'September',
    9: 'October',
    10: 'November',
    11: 'December'
}

console.log(`1. ${birthdayDate.toDateString()}`)  // 1. Mon Jun 29 1998
console.log(`2. ${birthdayDate.toLocaleDateString()}`)  // 2. 29/6/1998
console.log(`3. ${birthdayDate.toLocaleString()}`)  // 3. 29/6/1998, 9:25:00
console.log(`4. ${birthdayDate.getHours()}:${birthdayDate.getMinutes()}:${birthdayDate.getSeconds()}`)  // 4. 9:25:0
console.log(`5. ${birthdayDate.toLocaleTimeString()}`)  // 5. 9:25:00
console.log(`6. ${birthdayDate.getDay()}`)  // 6. 1
console.log(`7. ${weekDays[birthdayDate.getDay()]}`)  // 7. Monday
console.log(`8. ${birthdayDate.getMonth() + 1}`)  // 8. 6
console.log(`9. ${months[birthdayDate.getMonth()]}`)  // 9. June
console.log(`10. ${birthdayDate.toJSON()}`)  // 10. 1998-06-29T07:25:00.000Z

File: Roadmap/14 - FECHAS/python/nlarrea.py
"""
 * EJERCICIO:
 * Crea dos variables utilizando los objetos fecha (date, o semejante) de tu lenguaje:
 * - Una primera que represente la fecha (día, mes, año, hora, minuto, segundo) actual.
 * - Una segunda que represente tu fecha de nacimiento (te puedes inventar la hora).
 * Calcula cuántos años han transcurrido entre ambas fechas.
 *
 * DIFICULTAD EXTRA (opcional):
 * Utilizando la fecha de tu cumpleaños, formatéala y muestra su resultado de
 * 10 maneras diferentes. Por ejemplo:
 * - Día, mes y año.
 * - Hora, minuto y segundo.
 * - Día de año.
 * - Día de la semana.
 * - Nombre del mes.
 * (lo que se te ocurra...)
"""

from datetime import datetime, date

current_date = datetime.now().date()
bday_date = datetime.strptime("1998-06-29 09:25:00", "%Y-%m-%d %H:%M:%S")


def diffYears(date_one: date, date_two: date):
    return abs(date_one.year - date_two.year)


print(diffYears(current_date, bday_date))


"""
 * DIFICULTAD EXTRA (opcional):
 * Utilizando la fecha de tu cumpleaños, formatéala y muestra su resultado de
 * 10 maneras diferentes. Por ejemplo:
 * - Día, mes y año.
 * - Hora, minuto y segundo.
 * - Día de año.
 * - Día de la semana.
 * - Nombre del mes.
 * (lo que se te ocurra...)
"""

WEEK_DAYS = {
    0: "Monday",
    1: "Tuesday",
    2: "Wednesday",
    3: "Thursday",
    4: "Friday",
    5: "Saturday",
    6: "Sunday",
}

MONTH_NAMES = {
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December",
}

print(f"1. {bday_date.ctime()}")
print(f"2. {bday_date.date()}")
print(f"3. {bday_date.isoformat()}")
print(f"4. {bday_date.hour}:{bday_date.minute}:{bday_date.second}")
print(f"5. {bday_date.isocalendar()}")
print(f"6. {bday_date.day}")
print(f"7. {WEEK_DAYS[bday_date.weekday()]}")
print(f"8. {bday_date.month}")
print(f"9. {MONTH_NAMES[bday_date.month]}")
print(f"10. {bday_date.time()}")
